### PR TITLE
*: consolidate some settings related to the collation

### PR DESCRIPTION
**Pull request by bb7133**
### What is changed and how it works?

This PR aims to update the following collation-related features:

1. Restrict the settings of collation-related system variables `collation_connection`, `collation_server` and `collation_database`. Before this PR, they can be set to anything but now only valid collations are permitted, for example:

    Before this PR:
    ```
    tidb> set @@collation_connection='some_nonexist_collation';
    Query OK, 0 rows affected (0.00 sec)
    tidb> set names utf8mb4 collation some_nonexist_collation;
    Query OK, 0 rows affected (0.00 sec)
    ```

    After this PR(same with MySQL):
    ```
    tidb> set @@collation_connection='some_nonexist_collation';
    ERROR 1273 (HY000): Unknown collation: 'some_nonexist_collation'
    tidb> set names utf8mb4 collate some_nonexist_collation;
    ERROR 1273 (HY000): Unknown collation: 'some_nonexist_collation'
    ```

    **NOTICE**: This potentially breaks compatibility with old TiDB, but it should be fine since it means nothing if such system variables are set to an invalid value.

2. Forbid unsupported collations when new collation is enabled(see #14734 for more details about the switch):
    Before this PR when new collation is enabled:
    ```
    tidb> set names utf8mb4 collate utf8mb4_unicode_ci;
    Query OK, 0 rows affected (0.00 sec)
    ```
    After this PR when new collation is enabled:
    ```
    tidb> set names utf8mb4 collate utf8mb4_unicode_ci;
    ERROR 1273 (HY000): Unsupported collation when new collation is enabled: 'utf8mb4_unicode_ci'
    ```

3. Update the result of `information_schema`.`COLLATIONS` when new collation is enabled: `utf8mb4_general_ci` and `utf8_general_ci` are added since they're supported now:
    ```
    tidb> show collation;
    +--------------------+---------+------+---------+----------+---------+
    | Collation          | Charset | Id   | Default | Compiled | Sortlen |
    +--------------------+---------+------+---------+----------+---------+
    | binary             | binary  |   63 | Yes     | Yes      |       1 |
    | ascii_bin          | ascii   |   65 | Yes     | Yes      |       1 |
    | latin1_bin         | latin1  |   47 | Yes     | Yes      |       1 |
    | utf8mb4_bin        | utf8mb4 |   46 | Yes     | Yes      |       1 |
    | utf8_bin           | utf8    |   83 | Yes     | Yes      |       1 |
    | utf8mb4_general_ci | utf8mb4 |   45 |         | Yes      |       1 |
    | utf8_general_ci    | utf8    |   33 |         | Yes      |       1 |
    +--------------------+---------+------+---------+----------+---------+
    7 rows in set (0.00 sec)
    ```

### Related changes

- NA

### Check List 

Tests 

- Integration test
- Manual test (add detailed scripts or steps below)
    See above.

Side effects

- Breaking backward compatibility

### Release note

- When the new collation framework is enabled, collations not supported are now forbidden.
- Validate the value of system variables `collation_connection`, `collation_server` and `collation_database`,  `Unknown collation` will be returned if they're not valid.

### CODE REVIEW

**Comment by bb7133** [comment]
PTAL @wjhuang2016 @qw4990

File: ddl/serial_test.go [hunk 1]
```diff
@@ -910,3 +911,38 @@ func (s *testSerialSuite) TestAutoRandom(c *C) {
...
+	mustGetUnsupportedCollation("alter table t convert to charset utf8mb4 collate utf8mb4_unicode_ci", "utf8mb4_unicode_ci")
+
+	// Test collation of columns.
+	mustGetUnsupportedCollation("create table t1(a varchar(20) collate utf8mb4_unicode_ci", "utf8mb4_unicode_ci")
+	mustGetUnsupportedCollation("create table t1(a varchar(20) charset utf8 collate utf8_unicode_ci", "utf8_unicode_ci")
+	mustGetUnsupportedCollation("alter table t1 modify a varchar(20) collate utf8mb4_unicode_ci", "utf8mb4_unicode_ci")
+	mustGetUnsupportedCollation("alter table t1 modify a varchar(20) charset utf8 collate utf8_unicode_ci", "utf8_unicode_ci")
+	mustGetUnsupportedCollation("alter table t1 modify a varchar(20) charset utf8 collate utf8_unicode_ci", "utf8_unicode_ci")
+
+	// TODO(bb7133): fix the following cases by setting charset from collate firstly.
```
Comment (wjhuang2016) [inline ddl/serial_test.go]: Is this difficult to implement now?

Comment (bb7133) [reply ddl/serial_test.go]: No, it is easy...but it should be fixed in another PR.

Comment (bb7133) [reply ddl/serial_test.go]: No, it's easy. But I think we should fix it in another PR.

Comment (bb7133) [reply ddl/serial_test.go]: It's just about the 'correct' error message.

File: util/collate/collate.go [hunk 2]
```diff
@@ -111,62 +125,79 @@ func GetCollator(collate string) Collator {
...
 	return name
 }
 
-type binPaddingCollator struct {
+// GetCollationByName wraps charset.GetCollationByName, it checks the collation.
+func GetCollationByName(name string) (coll *charset.Collation, err error) {
+	if coll, err = charset.GetCollationByName(name); err != nil {
+		return nil, errors.Trace(err)
+	}
+	if atomic.LoadInt32(&newCollationEnabled) == 1 {
```
Comment (tangenta) [inline util/collate/collate.go]:
```suggestion
	if NewCollationEnabled() {
```

Same for all the `atomic.LoadInt32(&newCollationEnabled) == 1` in this file.

**Comment by tangenta** [comment]
Is it too late? :(

File: util/collate/collate.go (hunk 2 above)
Comment (bb7133) [reply util/collate/collate.go]: Got your point, but this code style is commonly used by collate.go, I think we can get another PR for them all.

**Comment by bb7133** [comment]
> Is it too late? :(

Not exactly, see my comment above :).

**Pull request merged**